File: implementations/typescript/packages/llmproxy/src/redirect.ts
// SPDX-License-Identifier: Apache-2.0
import {record} from "@psp-cdl/core";
import {evaluateBatch,type PolicyInput} from "@psp-cdl/cdl";
import {type Principal} from "@psp-cdl/api-server";
import {WorkflowStore,bounded,validRedirect,validRedirectTarget,REDIRECT_PROFILE,type AccessContext} from "@psp-cdl/api-server/persistence";
import {McpDispatchGate,bindingDigest} from "@psp-cdl/mcpproxy";
import {DurableLlmLoop,type DurableHost,type DurableOptions} from "./durable.js";
import {LoopError,type ProviderRegistration} from "./loop.js";

export {REDIRECT_PROFILE};
export interface RedirectHost extends DurableHost {
  resolveRedirect(principal:Principal,binding:Record<string,unknown>,target:string):
    {nodeId:string;nodeVersion:string;policyVersion:string;expiresAt:number}|Promise<{nodeId:string;nodeVersion:string;policyVersion:string;expiresAt:number}>;
  redirectPolicy(principal:Principal,binding:Record<string,unknown>,data:Record<string,unknown>):
    {bindingDigest:string;resources:PolicyInput[]}|Promise<{bindingDigest:string;resources:PolicyInput[]}>;
}
const fail=(code:string):never=>{throw new LoopError(code);};
function copy(value:unknown,code="INVALID_REDIRECT"):any {try{return bounded(value);}catch{return fail(code);}}
async function call<T>(fn:()=>T|Promise<T>):Promise<T> {try{return await fn();}catch{return fail("HOST_ERROR");}}

/** Host-selected, same-owner completion handoff. Never invokes the target provider. */
export class RedirectingLlmLoop extends DurableLlmLoop {
  private readonly target:string;
  protected override get completionPolicy():string {return "redirect";}
  constructor(store:WorkflowStore,gate:McpDispatchGate,private readonly redirectHost:RedirectHost,provider:ProviderRegistration,configuration:{postCompletion:"redirect";target:string}) {
    super(store,gate,redirectHost,provider,{postCompletion:"lockdown"});
    if(!store.redirectTurns||[redirectHost.resolveRedirect,redirectHost.redirectPolicy].some(f=>typeof f!=="function"))fail("INVALID_CONFIGURATION");
    if(!record(configuration)||Object.keys(configuration).sort().join(",")!=="postCompletion,target"||configuration.postCompletion!=="redirect"||!validRedirectTarget(configuration.target))fail("UNSUPPORTED_POST_COMPLETION");
    this.target=configuration.target;
  }
  protected override async prepareCommit(p:Principal,binding:Record<string,unknown>,command:Record<string,unknown>,options:DurableOptions):Promise<Record<string,unknown>> {
    this.check(options);
    let redirect:Record<string,unknown>|null=null;
    if(command.complete) {
      const resolved=copy(await call(()=>this.redirectHost.resolveRedirect(copy(p),copy({...binding,requestId:command.requestId,postCompletion:"redirect"}),this.target)));
      if(!record(resolved)||Object.keys(resolved).sort().join(",")!=="expiresAt,nodeId,nodeVersion,policyVersion")fail("INVALID_REDIRECT");
      redirect={...resolved,target:this.target};
      const source=await this.store.execute({tenantId:p.tenantId,subjectId:p.subjectId},{action:"getSession",sessionId:command.sessionId});
      if(!validRedirect(redirect)||redirect.nodeId===command.nodeId||(redirect.expiresAt as number)<=this.host.now()||(redirect.expiresAt as number)>(source.expiresAt as number))fail("INVALID_REDIRECT");
    }
    this.check(options);
    return {...command,action:"commitRedirectTurn",redirect};
  }
  protected override async authorizeCommit(p:Principal,binding:Record<string,unknown>,context:AccessContext,options:DurableOptions):Promise<void> {
    await super.authorizeCommit(p,binding,context,options);
    if(!context.command.complete)return;
    const bound={...binding,requestId:context.command.requestId,postCompletion:"redirect",redirect:context.result.redirect,commandDigest:bindingDigest(context.command)};
    const data={output:context.command.output,retained:context.command.retained};
    const policy=copy(await call(()=>this.redirectHost.redirectPolicy(copy(p),copy(bound),copy(data))),"INVALID_POLICY");
    if(!record(policy)||Object.keys(policy).sort().join(",")!=="bindingDigest,resources"||policy.bindingDigest!==bindingDigest(bound)||!Array.isArray(policy.resources)||!policy.resources.length)fail("INVALID_POLICY");
    const decision=evaluateBatch(policy.resources).decision;
    if(decision==="unsupported")fail("UNSUPPORTED_POLICY");
    if(decision!=="allow")fail("REDIRECT_DENIED");
    this.check(options,(context.result.redirect as Record<string,number>).expiresAt);
  }
  protected override present(receipt:Record<string,unknown>,recovered:boolean):Record<string,unknown> {
    return {...super.present(receipt,recovered),...(receipt.redirect?{redirect:copy(receipt.redirect)}:{})};
  }
}

File: implementations/python/packages/llmproxy/src/psp_cdl_llmproxy/redirect.py
# SPDX-License-Identifier: Apache-2.0
"""Host-selected, same-owner completion handoff; no automatic target inference."""
from psp_cdl_cdl import evaluate_batch
from psp_cdl_api_server.redirect_state import REDIRECT_PROFILE, valid_redirect, valid_redirect_target
from psp_cdl_mcpproxy import binding_digest
from .durable import DurableLlmLoop, call, actor
from .loop import LoopError, copy


class RedirectingLlmLoop(DurableLlmLoop):
    _completion_policy = "redirect"

    def __init__(self,store,gate,host,provider,configuration=None):
        super().__init__(store,gate,host,provider,{"postCompletion":"lockdown"})
        if not store.redirect_turns or any(not callable(getattr(host,k,None)) for k in ("resolve_redirect","redirect_policy")): raise LoopError("INVALID_CONFIGURATION")
        if type(configuration) is not dict or set(configuration)!={"postCompletion","target"} or configuration["postCompletion"]!="redirect" or not valid_redirect_target(configuration["target"]): raise LoopError("UNSUPPORTED_POST_COMPLETION")
        self._target=configuration["target"]

    def _prepare_commit(self,p,binding,command,options):
        self._check(options)
        redirect=None
        if command["complete"]:
            resolved=copy(call(lambda:self._host.resolve_redirect(copy(p),copy({**binding,"requestId":command["requestId"],"postCompletion":"redirect"}),self._target)),"INVALID_REDIRECT")
            if type(resolved) is not dict or set(resolved)!={"expiresAt","nodeId","nodeVersion","policyVersion"}: raise LoopError("INVALID_REDIRECT")
            redirect={**resolved,"target":self._target}
            source=self._store.execute(actor(p),{"action":"getSession","sessionId":command["sessionId"]})
            if not valid_redirect(redirect) or redirect["nodeId"]==command["nodeId"] or not self._host.now()<redirect["expiresAt"]<=source["expiresAt"]: raise LoopError("INVALID_REDIRECT")
        self._check(options)
        return {**command,"action":"commitRedirectTurn","redirect":redirect}

    def _authorize_commit(self,p,binding,context,options):
        super()._authorize_commit(p,binding,context,options)
        command=context["command"]
        if not command["complete"]: return
        bound={**binding,"requestId":command["requestId"],"postCompletion":"redirect","redirect":context["result"]["redirect"],"commandDigest":binding_digest(command)}
        data={"output":command["output"],"retained":command["retained"]}
        policy=copy(call(lambda:self._host.redirect_policy(copy(p),copy(bound),copy(data))),"INVALID_POLICY")
        if type(policy) is not dict or set(policy)!={"bindingDigest","resources"} or policy["bindingDigest"]!=binding_digest(bound) or type(policy["resources"]) is not list or not policy["resources"]: raise LoopError("INVALID_POLICY")
        decision=evaluate_batch(policy["resources"])["decision"]
        if decision=="unsupported": raise LoopError("UNSUPPORTED_POLICY")
        if decision!="allow": raise LoopError("REDIRECT_DENIED")
        self._check(options,context["result"]["redirect"]["expiresAt"])

    @staticmethod
    def _present(receipt,recovered):
        return {**DurableLlmLoop._present(receipt,recovered),**({"redirect":copy(receipt["redirect"])} if "redirect" in receipt else {})}
